hal/src/nRF52840/ble_hal.cpp

```diff
@@ -2325,24 +2456,40 @@ int BleObject::ConnectionsManager::processSecurityEventFromThread(const ble_evt_
                 linkEvent.conn_handle = connection->info.conn_handle;
                 notifyLinkEvent(linkEvent);
             }
-            // User application may call rejectPairing() to update the pairing state in event handler
-            if (connection->pairState != BLE_PAIRING_STATE_REJECTED) {
-                ble_gap_sec_params_t secParams = {};
-                if (ioCaps_ != BLE_IO_CAPS_NONE) {
-                    secParams.mitm = true;
-                }
-                secParams.io_caps        = toPlatformIoCaps(ioCaps_);
-                secParams.oob            = false;
-                secParams.min_key_size   = BLE_ENC_MIN_KEY_SIZE;
-                secParams.max_key_size   = BLE_ENC_MAX_KEY_SIZE;
-                secParams.kdist_own.enc  = true;
-                secParams.kdist_own.id   = true;
-                secParams.kdist_peer.enc = true;
-                secParams.kdist_peer.id  = true;
-                ret = sd_ble_gap_sec_params_reply(event->evt.gap_evt.conn_handle, BLE_GAP_SEC_STATUS_SUCCESS, &secParams, nullptr);
+        }
+        // User application may call rejectPairing() to update the pairing state in event handler
+        if (connection->pairState == BLE_PAIRING_STATE_REJECTED) {
+            return ret;
+        }
+        ble_gap_sec_params_t secParams = {};
+        ble_gap_sec_params_t* pSecParams = nullptr; // For central role, default it to nullptr, since we have set the secure parameters in startPairing().
+        ble_gap_sec_keyset_t keyset = {};
+        ble_gap_sec_keyset_t* pKeyset = nullptr; // nullptr for BLE legacy pairing
+        bool reject = false;
+        if (pairingConfig_.algorithm == BLE_PAIRING_ALGORITHM_LESC_ONLY) {
+            if (peerSecParams.lesc == false) {
+                reject = true;
+            }
+        }
+        if (!reject && pairingConfig_.algorithm != BLE_PAIRING_ALGORITHM_LEGACY_ONLY ) {
+            if (publicKeysPrepare(connection) != SYSTEM_ERROR_NONE) {
+                LOG(ERROR, "Failed to allocate memory for public keys.");
+                reject = true;
+            } else {
+                keyset.keys_own.p_pk = localPublicKey_.get();
+                keyset.keys_peer.p_pk = connection->peerPublicKey.get();
+                pKeyset = &keyset;
             }
         }
+        if (!reject && connection->info.role == BLE_ROLE_PERIPHERAL) {
+            initSecParams(pairingConfig_, secParams);
+            pSecParams = &secParams;
+        }
+        ret = sd_ble_gap_sec_params_reply(event->evt.gap_evt.conn_handle, reject ? BLE_GAP_SEC_STATUS_PAIRING_NOT_SUPP : BLE_GAP_SEC_STATUS_SUCCESS, pSecParams, pKeyset);
         if (ret != NRF_SUCCESS) {
```

Comment:
Or if `rejected == true` ?

Reply:
`ret == NRF_SUCCESS` will result `BLE_GAP_EVT_AUTH_STATUS` event and followed by doing something properly as well.